miner/src/block_assembler.rs

```diff
@@ -324,8 +374,13 @@ impl<CS: ChainStore + 'static> BlockAssembler<CS> {
         // bytes, they really serve the same purpose at the moment
         let block_reward = self.shared.block_reward(header.number() + 1);
         let mut fee = 0;
-        for pe in pes {
-            fee += self.shared.calculate_transaction_fee(&pe.transaction)?;
+        for (i, pe) in pes.iter().enumerate() {
+            // depends cells may produced from previous tx
+            let fee_calculator = FeeCalculator {
```

Comment:
It's better to use `HashMap` in FeeCalculator as a previously calculated txs capacity cache, and change the `calculate_transaction_fee` function to mut, which will simplify code here.